build/scripts/script.js

```diff
@@ -519,84 +525,153 @@ function updateSearchResults(value) {
       '<div class="message">' + "No result for that search." + "</div>";
     return;
   }
+
+  // only consider the first 30 results
+  const searchResultsSliced = searchResults.slice(0, 29);
+  const searchResultSorted = reorderSearchResults(searchResultsSliced);
   searchResultsElt.innerHTML = "";
 
-  for (let resIdx = 0; resIdx < searchResults.length && resIdx < 30; resIdx++) {
-    const res = searchResults[resIdx];
-    const links = searchIndexLinks[+res.refIndex];
-    const contentDiv = document.createElement("div");
-    contentDiv.className = "search-result-item";
-
-    const locationDiv = document.createElement("div");
-    locationDiv.className = "search-result-location";
-
-    let needSeparator = false;
-    if (res.item.h1 !== undefined && res.item.h1 !== "") {
-      let linkH1;
-      if (links.anchorH1 !== undefined) {
-        const href = rootUrl + "/" + links.file + "#" + links.anchorH1;
-        linkH1 = document.createElement("a");
-        linkH1.href = href;
-      } else {
-        linkH1 = document.createElement("span");
-      }
-      linkH1.className = "h1";
-      linkH1.textContent = res.item.h1;
-      locationDiv.appendChild(linkH1);
-      needSeparator = true;
+  let previousItem = null;
+  for (const searchResult of searchResultSorted) {
+    let displayFromLevel =
+      previousItem === null
+        ? 0
+        : getCommonAncestorLevel(searchResult, previousItem);
+    previousItem = searchResult;
+    const elems = createHeadingElements(searchResult, displayFromLevel);
+    elems.forEach((elem) => searchResultsElt.appendChild(elem));
+  }
+}
+
+/**
+ * Compares two search result items to determine the deepest common ancestor level.
+ *
+ * @param {Object} itemA - The first search result item.
+ * @param {Object} itemB - The second search result item.
+ * @returns {number} - Returns 0 if items does not share same file, 1 if h1 is different, etc...
+ */
+function getCommonAncestorLevel(searchResultItemA, searchResultItemB) {
+  if (searchResultItemA.file !== searchResultItemB.file) {
+    return 0;
+  }
+
+  if (searchResultItemA.h1 !== searchResultItemB.h1) {
+    return 1;
+  }
+
+  if (searchResultItemA.h2 !== searchResultItemB.h2) {
+    return 2;
+  }
+
+  if (searchResultItemA.h3 !== searchResultItemB.h3) {
+    return 3;
+  }
+  return 3;
```

Comment:
So there's no difference if they share or not their h3? (Could be just a different occurence of the body?)

Reply:
Yes the condition is not necessary, I can return 3 in every case at this point.